File: Assignment 6/6.py
import random
import time

# Function to check whether the list is in sorted order or not
# VERIFICATION
def increasing(l):
    for i in range(0,len(l)-1):
        if(l[i+1]<l[i]):
            return False
    return True

# Defining a list of n size with random value ranging between 0 to 101
n = 8
lt = []
for i in range(n):
    lt.append(random.randint(1,100))

# Finding Time required by Deterministic sorting algorithm
start_time1 = time.time()
sorted_list = sorted(lt)
end_time1 = time.time()

# Finding Time required by Non-Deterministic sorting algorithm
start_time2 = time.time()
while True:
    nd_sorted_list = []
    lt_len = len(lt)
    visited = {}
    i = 0
    while(i < lt_len):
        index = random.randint(0,lt_len-1)
        if index not in visited.keys():
            visited[index] = 1
            nd_sorted_list.append(lt[index]) 
            i = i+1 
    if(increasing(nd_sorted_list)):
        break
end_time2 = time.time()

print(lt)
print(nd_sorted_list)


print("Non Deterministic Sort takes: "+str(end_time2-start_time2)+"\nWhile Deterministic Sort takes: "+str(end_time1-start_time1))

File: Assignment 6/6.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

class Sorting {

    public static boolean increasing(List<Integer> arr) {
        for(int i = 0; i<arr.size() - 2; i++) {
            if(arr.get(i+1) < arr.get(i)) return false;
        }
        return true;
    }

    public static void nonDeterministicSorting(List<Integer> arr) {
        long time = System.currentTimeMillis();
        Random r = new Random();
        List<Integer> sorted = new ArrayList<>();
        Map<Integer, Integer> visited = new HashMap<>();
        int count = 0;
        while(count < 20) {
            for(int i = 0; i < arr.size(); i++) {
                int index = r.nextInt(arr.size() - 1);
                if(!visited.containsKey(index)) {
                    visited.put(index, 1);
                    sorted.add(arr.get(index));
                }
            }
            if(increasing(arr)) break;
            count++;
        }
        time = System.currentTimeMillis() - time;
        // System.out.println(sorted);
        System.out.println("Non Deterministic time takes " + time + "ms");
    }

    public static void deterministicSorting(List<Integer> arr) {
        long time = System.currentTimeMillis();
        Collections.sort(arr);
        time = System.currentTimeMillis() - time;
        // System.out.println(arr);
        System.out.println("Deterministic time takes " + time + "ms");
    }

    public static void main(String[] args) {
        List<Integer> l = new ArrayList<>();
        List<Integer> l1 = new ArrayList<>();
        Random r = new Random();
        for(int i = 0; i< 100; i++) {
            l.add(r.nextInt(100));
            l1.add(r.nextInt(100));
        }
        deterministicSorting(l);
        nonDeterministicSorting(l1);
    }
}